config/module/versions_test.go

```diff
@@ -58,3 +58,29 @@ func TestNewestInvalidModuleVersion(t *testing.T) {
 		t.Fatalf("expected version %q, got %q", expected, m.Version)
 	}
 }
+
+func TestNewestModulesWithMetadata(t *testing.T) {
+	mpv := &response.ModuleProviderVersions{
+		Source: "registry/test/module",
+		Versions: []*response.ModuleVersion{
+			{Version: "0.9.0"},
+			{Version: "0.9.0+def"},
+			{Version: "0.9.0+abc"},
+			{Version: "0.9.0+xyz"},
+		},
+	}
+
+	// with metadata and explicit version request
+	expected := "0.9.0+def"
+	m, _ := newestVersion(mpv.Versions, "=0.9.0+def")
+	if m.Version != expected {
+		t.Fatalf("expected version %q, got %q", expected, m.Version)
+	}
+
+	// respect explicit equality, but >/</~ will be ignored
+	expected = "0.9.0"
+	m, _ = newestVersion(mpv.Versions, "~>0.9.0+abc")
```

Comment:
Including a build metadata string in a non-exact version constraint seems like an odd thing to do, and I suspect a user would be surprised to see the result being tested for here of actually selecting `0.9.0` in that case.

Do you think we should instead make this return an error, stating explicitly that build metadata can only be used in exact version constraints? That would hopefully guide the user towards writing something with a more predictable result.